rust/src/operations/update.rs

```diff
@@ -0,0 +1,921 @@
+        let c = "__delta_rs_".to_string() + &column.name;
+        expressions.push((predicate_expr, c.clone()));
+        control_columns.insert(c);
+    }
+
+    let projection_update: Arc<dyn ExecutionPlan> =
+        Arc::new(ProjectionExec::try_new(expressions, count_plan.clone())?);
+
+    // Project again to remove __delta_rs columns and rename update columns to their original name
+    let mut expressions: Vec<(Arc<dyn PhysicalExpr>, String)> = Vec::new();
+    let scan_schema = projection_update.schema();
+    for (i, field) in scan_schema.fields().into_iter().enumerate() {
+        if !control_columns.contains(field.name()) {
+            match map.get(field.name()) {
+                Some(value) => {
+                    expressions.push((
+                        Arc::new(expressions::Column::new(field.name(), *value)),
+                        field.name().to_owned(),
+                    ));
+                }
+                None => {
+                    expressions.push((
+                        Arc::new(expressions::Column::new(field.name(), i)),
+                        field.name().to_owned(),
+                    ));
+                }
+            }
+        }
+    }
+
+    let projection: Arc<dyn ExecutionPlan> = Arc::new(ProjectionExec::try_new(
+        expressions,
+        projection_update.clone(),
+    )?);
+
+    let add_actions = write_execution_plan(
+        snapshot,
+        state.clone(),
+        projection.clone(),
+        table_partition_cols.clone(),
+        object_store.clone(),
+        Some(snapshot.table_config().target_file_size() as usize),
+        None,
+        writer_properties,
+        &cast_options,
+    )
+    .await?;
+
+    let count_metrics = count_plan.metrics().unwrap();
+
+    metrics.num_updated_rows = count_metrics
+        .sum_by_name("num_updated_rows")
+        .unwrap()
+        .as_usize();
+    metrics.num_copied_rows = count_metrics
+        .sum_by_name("num_copied_rows")
+        .unwrap()
+        .as_usize();
+
+    let deletion_timestamp = SystemTime::now()
+        .duration_since(UNIX_EPOCH)
+        .unwrap()
+        .as_millis() as i64;
+    let mut actions: Vec<Action> = add_actions.into_iter().map(Action::add).collect();
+
+    metrics.num_added_files = actions.len();
+    metrics.num_removed_files = candidates.candidates.len();
+
+    for action in candidates.candidates {
+        actions.push(Action::remove(Remove {
+            path: action.path,
+            deletion_timestamp: Some(deletion_timestamp),
+            data_change: true,
+            extended_file_metadata: Some(true),
+            partition_values: Some(action.partition_values),
+            size: Some(action.size),
+            tags: None,
+        }))
+    }
+
+    metrics.execution_time_ms = Instant::now().duration_since(exec_start).as_micros();
+
+    // Do not make a commit when there are zero updates to the state
+    if !actions.is_empty() {
```

Comment:
Shouldn't we have eliminating this possibility when finding files?

Reply:
Yes good point.  I made a change to return early when find files determines zero candidates. Added a test to ensure the metrics are correct too.